examples/megssm/models.py

```diff
@@ -0,0 +1,872 @@
+                                   % (filename,))
+        for name in param_names:
+            if name == 'log_sigsq_lst':
+                self.log_sigsq_lst = [l.copy() for l in loaddict[name]]
+            elif name in ('lam0', 'lam1'):
+                self.__setattr__(name, float(loaddict[name]))
+            else:
+                self.__setattr__(name, loaddict[name].copy())
+
+        # return remaining saved items, if there are any
+        others = {key : loaddict[key] for key in loaddict.keys() \
+                                      if key not in param_names}
+        if len(others.keys()) > 0:
+            return others
+
+    @staticmethod
+    def R_(snsr_cov, fwd_src_snsr, sigsq_vals, which_roi):
+        return snsr_cov + np.dot(fwd_src_snsr, sigsq_vals[which_roi][:,None]*fwd_src_snsr.T)
+
+    def L2_obj(self, At, L_roi_cov):
+        
+        # import autograd.numpy
+        # if isinstance(At,autograd.numpy.numpy_boxes.ArrayBox):
+        #     At = At._value
+            
+        AtB2T = einsum2('tik,tjk->tij', At, self._B2)
+        B2AtT = einsum2('tik,tjk->tij', self._B2, At)
+        tmp = einsum2('tik,tkl->til', At, self._B3)
+        AtB3AtT = einsum2('til,tjl->tij', tmp, At)
+        elbo_2 = np.sum(self._B1 - AtB2T - B2AtT + AtB3AtT, axis=0)
+
+        L_roi_cov_inv_elbo_2 = solve_triangular(L_roi_cov, elbo_2, lower=True)
+        obj = np.trace(solve_triangular(L_roi_cov, L_roi_cov_inv_elbo_2, lower=True,
+                                        trans='T'))
+        obj = obj / self._ntrials_all
+
+        if self._penalty == 'ridge':
+            obj += self.lam0*np.sum(At**2)
+        AtmAtm1_2 = (At[1:] - At[:-1])**2
+        obj += self.lam1*np.sum(AtmAtm1_2)
+
+        return obj
+
+    # TODO: convert to instance method
+    @staticmethod
+    def L3_obj(log_sigsq_vals, snsr_cov, fwd_src_snsr, which_roi, B4, ntrials, timepts):
+        R = MEGLDS.R_(snsr_cov, fwd_src_snsr, np.exp(log_sigsq_vals), which_roi)
+        try:
+            L_R = np.linalg.cholesky(R)
+        except LinAlgError:
+            return np.finfo('float').max
+        L_R_inv_B4 = solve_triangular(L_R, B4, lower=True)
+        return (ntrials*timepts*2.*np.sum(np.log(np.diag(L_R)))
+                + np.trace(solve_triangular(L_R, L_R_inv_B4, lower=True,
+                                            trans='T')))
+
+
+    @property
+    def A(self):
+        return self._A
+
+    @A.setter
+    def A(self, A):
+        self._A = A
+
+    @property
+    def roi_cov(self):
+        return self._roi_cov
+
+    @roi_cov.setter
+    def roi_cov(self, roi_cov):
+        self._roi_cov = roi_cov
+
+    @property
+    def mu0(self):
+        return self._mu0
+
+    @mu0.setter
+    def mu0(self, mu0):
+        self._mu0 = mu0
+
+    @property
+    def roi_cov_0(self):
+        return self._roi_cov_0
+
+    @roi_cov_0.setter
+    def roi_cov_0(self, roi_cov_0):
+        self._roi_cov_0 = roi_cov_0
+
+    @property
+    def log_sigsq_lst(self):
```

Comment:
All of this stuff looks like it should be private to me. That also means we don't need a `@property` or a `@whatever.setter` for each of them. We just use private attributes directly

Reply:
So I should be able to remove all of these `@property` and `@some.setter` lines right? Does that mean there needs to be additional checks on variable format somewhere? I don't think any of these have any checks as they are written here. I'd also like to talk about which vars and functions to make private. In my mind, all of the vars and functions in `models.py` should be private.